desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,1349 @@
+                legend[label] = line
+
+    fig, ax = plt.subplots()
+    for knot in B.x:
+        add(ax.axvline(x=knot, color="tab:blue", alpha=0.25, label="knot"))
+    z = jnp.linspace(
+        start=B.x[0] if start is None else start,
+        stop=B.x[-1] if stop is None else stop,
+        num=num,
+    )
+    add(ax.plot(z, B(z), label=r"$\vert B \vert (\zeta)$"))
+
+    if pitch is not None:
+        b = 1 / jnp.atleast_1d(pitch)
+        for val in b:
+            add(ax.axhline(val, color="tab:purple", alpha=0.25, label=r"$1 / \lambda$"))
+        bp1, bp2 = jnp.atleast_2d(bp1, bp2)
+        for i in range(bp1.shape[0]):
+            bp1_i, bp2_i = map(_filter_not_nan, (bp1[i], bp2[i]))
+            add(
+                ax.scatter(
+                    bp1_i,
+                    jnp.full_like(bp1_i, b[i]),
+                    marker="v",
+                    color="tab:red",
+                    label="bp1",
+                )
+            )
+            add(
+                ax.scatter(
+                    bp2_i,
+                    jnp.full_like(bp2_i, b[i]),
+                    marker="^",
+                    color="tab:green",
+                    label="bp2",
+                )
+            )
+
+    ax.set_xlabel(r"Field line $\zeta$")
+    ax.set_ylabel(r"$\vert B \vert \sim 1 / \lambda$")
+    ax.legend(legend.values(), legend.keys())
+    if id is not None:
+        title = f"{title}. id = {id}."
+    ax.set_title(title)
+    if show:
+        plt.tight_layout()
+        plt.show()
+        plt.close()
+    return fig, ax
+
+
+def affine_bijection(x, a, b):
+    """[−1, 1] ∋ x ↦ y ∈ [a, b]."""
+    y = (x + 1) / 2 * (b - a) + a
+    return y
+
+
+def grad_affine_bijection(a, b):
+    """Gradient of affine bijection."""
+    dy_dx = (b - a) / 2
+    return dy_dx
+
+
+def automorphism_arcsin(x):
+    """[-1, 1] ∋ x ↦ y ∈ [−1, 1].
+
+    The gradient of the arcsin automorphism introduces a singularity that augments
+    the singularity in the bounce integral. Therefore, the quadrature scheme
+    used to evaluate the integral must work well on functions with large
+    derivative near the boundary.
+
+    The arcsin automorphism pulls points in [−1, 1] away from the boundary.
+    This can reduce floating point error if paired with a quadrature
+    scheme that is aggressive with placing nodes near endpoints, such as
+    Tanh-Sinh quadrature.
+
+    Parameters
+    ----------
+    x : Array
+        Points to transform.
+
+    Returns
+    -------
+    y : Array
+        Transformed points.
+
+    """
+    y = 2 * jnp.arcsin(x) / jnp.pi
+    return y
+
+
+def grad_automorphism_arcsin(x):
+    """Gradient of arcsin automorphism."""
+    dy_dx = 2 / (jnp.sqrt(1 - x**2) * jnp.pi)
+    return dy_dx
+
+
+grad_automorphism_arcsin.__doc__ += "\n" + automorphism_arcsin.__doc__
+
+
+def automorphism_sin(x, s=0, m=10):
+    """[-1, 1] ∋ x ↦ y ∈ [−1, 1].
+
+    The gradient of the sin automorphism is Lipschitz.
```

Comment:
Do you mean Lipschitz continuous? What is a Lipschitz gradient?

Reply:
Here is the definition
definition 12.6 (Lipschitzness) Let $C \subset R^d$ . A function $f  \colon R^d \to R^k$ is
ρ-Lipschitz over $C$ if for every $w_1 , w_2 \in C$ we have that $|| f (w_1) − f (w_2) || \leq \rho || w_1 − w_2 ||$.

I am saying that our automorphism, let's call it $a$, has a gradient $\nabla a$, which is Lipschitz. Lipschitz continuous is the same thing as Lipshitz. A change of variables with Lipshitz gradient cannot introduce singularities